### DISCUSSION TITLE
Breakpoints in Microsoft Visual Studio no longer work after migrating from Nunit2 to Nunit3 with TestCentric.GuiRunner.1.6.4

### DISCUSSION THREAD

**Discussion by ErikBee** [TestCentric/testcentric-gui#1112] [Q&A] [answered]
When starting the application from MVS debug with testcentric.exe my.dll, the breakpoints are disabled.
I assume this is caused by the fact that a testcentric-agent.exe is started.
I can manually attach to testcentric-agent.exe and debug again, but this is quite inconvenient.

Is there an alternative way, so no manual attach is needed?
(e.g. would it be possible to start the agent directly?)

**Comment by CharliePoole** [comment, marked as answer]
No, this is not available in 1.6.4. As a Workaround, create a source file similar to this in your test project:

```
#if DEBUG
using System.Diagnostics;
using NUnit.Framework;

[SetUpFixture]
public class EnableDebugging
{
    [OneTimeSetUp]
    public void EnableDebugger()
    { 
        if (!Debugger.IsAttached)
            Debugger.Launch();
    }
}
#endif
```

I used `#if DEBUG` because you are debugging and also to avoid this code ever being released to your users. Note that this code specifies no namespace, so the class is in the global namespace. The EnableDebugger method will run before any tests in your assembly are executed, launching the debugger on your behalf.

The run is paused at the Debugger.Launch statement. You can use that time to set breakpoints if you have not set them in advance. Continue execution and your breakpoints will be hit.

You may want to consider using the 2.0.0-beta5 release in lieu of 1.6.4. It's not fully production-ready so it's usefulness to you will depend on what features you need. FWIW, that beta does allow running an agent directly but the feature is only used internally for my own tests and the details are not yet documented. If you decide to try that approach, contact me. However, I think the above workaround is probably the best approach for now.

When your tests

**Reply by ErikBee** [reply]
Based on this proposal I now have a solution to automatic attach without popup.

```
using System.Runtime.InteropServices;
using EnvDTE80;

#if DEBUG
[SetUpFixture]
public class EnableDebugging
{
    public static DTE2 Attach(string processName = "")
    {
        DTE2 dte = (DTE2) Marshal.GetActiveObject("VisualStudio.DTE.17.0"); 
        var processes = dte.Debugger.LocalProcesses;
        foreach (var proc in processes.Cast<EnvDTE.Process>().Where(proc => proc.Name.IndexOf(processName) != -1))
        {
            proc.Attach();
        }
        return dte;
    }
}
#endif

namespace MyNameSpace
{
    public class MainClass
    {
#if DEBUG
        // Attach to testcentric agent, only one of the two will exist
        public static DTE2 dte_testcentric_1_6_4 = EnableDebugging.Attach("testcentric-agent.exe");
        public static DTE2 dte_testcentric_2_2_0_beta5 = EnableDebugging.Attach("net462-agent.exe");
#endif

```

Thanks for the support.
